firewood/src/merkle/stream.rs

```diff
@@ -7,399 +7,471 @@ use crate::{
     shale::{DiskAddress, ShaleStore},
     v2::api,
 };
-use futures::{stream::FusedStream, Stream};
-use helper_types::{Either, MustUse};
+use futures::{stream::FusedStream, Stream, StreamExt};
 use std::task::Poll;
+use std::{cmp::Ordering, iter::once};
 
 type Key = Box<[u8]>;
 type Value = Vec<u8>;
 
-enum IteratorState<'a> {
-    /// Start iterating at the specified key
-    StartAtKey(Key),
-    /// Continue iterating after the last node in the `visited_node_path`
-    Iterating {
-        check_child_nibble: bool,
-        visited_node_path: Vec<(NodeObjRef<'a>, u8)>,
+/// Represents an ongoing iteration over a node and its children.
+enum IterationNode<'a> {
+    // This node has not been returned yet.
+    Unvisited {
+        /// The key (as nibbles) of this node.
+        key: Key,
+        node: NodeObjRef<'a>,
+    },
+    // This node has been returned. Track which child to visit next.
```

Comment:
I think you can also use triple `/` here too
```suggestion
    /// This node has not been returned yet.
    Unvisited {
        /// The key (as nibbles) of this node.
        key: Key,
        node: NodeObjRef<'a>,
    },
    /// This node has been returned. Track which child to visit next.
```